### PR TITLE
add initial param cast resolution in dyno

### PR DESCRIPTION
**Pull request by arezaii**
This PR enables explicit casts from `param`s in dyno.

I needed the ability to get an `Immediate` for each of the `Param` types in dyno. Production handles this with an `immediate` field on the `VarSymbol`. In Dyno we already had the conditional logic to process a `Param` into the right `Immediate` with the value populated, so here I have added a helper function that either gets the value of the `param` if it's passed, or a default initialized value for the `Immediate` if the `Param* p` argument is `nullptr`. 

When we don't have a `Param` and just need an empty `Immediate` setup for a given `Type`, I created a helper to return a `paramtag` based on the `Type` so we can still use all the same conditional logic in `paramToImmediate`.

With this update, the demo program in the original issue (https://github.com/Cray/chapel-private/issues/5922) resolves using `testInteractive`.

TESTING:

- [x] added cast tests as given in reproducer in 5922

[reviewed by @mppf - thank you!]

### CODE REVIEW

File: frontend/lib/resolution/resolution-queries.cpp [hunk 1]
```diff
@@ -2851,6 +2851,29 @@ static bool resolveFnCallSpecial(Context* context,
   // TODO: .borrow()
   // TODO: chpl__coerceCopy
 
+  // explicit param casts are resolved here
+  if (ci.name() == USTR(":")) {
+    auto lhs = ci.actual(0).type();
+    auto rhs = ci.actual(1).type();
```
Comment (mppf) [inline frontend/lib/resolution/resolution-queries.cpp:2857]: IMO lhs and rhs are a bit unintuitive when talking about a cast. I'd prefer `fromValue` and `toType` or just `from` and `to`.

File: frontend/lib/types/Param.cpp [hunk 2]
```diff
@@ -89,157 +90,168 @@ bool Param::isParamOpFoldable(chpl::uast::PrimitiveTag op) {
...
-          ret.v_complex128.r = v.re;
-          ret.v_complex128.i = v.im;
-        } else {
-          CHPL_ASSERT(false && "case not handled");
-        }
-        return ret;
-      }
-    case paramtags::EnumParam:
-      {
+  if (t->isBoolType()) {
```
Comment (mppf) [inline frontend/lib/types/Param.cpp:129]: I don't think we should assume here that if `t` is a bool type that `p` is a bool param. So it seems to me that the switch was better (at least, the blocks here should perhaps be based on the param type first? I would believe you might have run into a case where the type `t` needs to be used that wasn't working yet.

File: frontend/lib/resolution/resolution-queries.cpp [hunk 3]
```diff
@@ -2868,10 +2868,40 @@ static bool resolveFnCallSpecial(Context* context,
...
+  if (ci.isOpCall() && ci.name() == USTR(":")) {
+    auto src = ci.actual(0).type();
+    auto dst = ci.actual(1).type();
+
+    bool isDstType = dst.kind() == QualifiedType::TYPE;
+    bool isParamTypeCast = src.kind() == QualifiedType::PARAM && isDstType;
+
+    if (isParamTypeCast) {
+        exprTypeOut = Param::fold(context, uast::PrimitiveTag::PRIM_CAST,
+                                  src, dst);
```
Comment (mppf) [inline frontend/lib/resolution/resolution-queries.cpp:2884]: There are some param casts that aren't legal, right? I think? For example, `myEnum.red: nothing` ? I suspect that most are legal but we'll need some checking that it's a valid cast somewhere & I don't think we have that yet. I'd expect it'd be better to check it here than in `Param::fold` but either way could work.

**Review by mppf — APPROVED** [review]

File: frontend/lib/resolution/resolution-queries.cpp (hunk 3 above)
Comment (arezaii) [reply frontend/lib/resolution/resolution-queries.cpp:2914]: I added the rules to catch illegal casts involving enum to nothing, abstract enum to int, and int to abstract enum. I found those in prod from `functionResolution.cpp`, although there may be others defined in other places

**Pull request merged**